### ISSUE TITLE
Multi-step prediction - regression.

### ISSUE THREAD

**Issue by andrewczgithub** [cornellius-gp/gpytorch#695]
Hi All!
Love the project.
I am just wondering if some these models can be used for **mult-step prediction** in the future.
Ie propagate future values.

I understand the prediction function as to gain a single point and output another point or can it propagate multiple points, specified by the user?

Many thanks,
Best,
Andrew

**Comment by jacobrgardner** [comment]
@andrewczgithub I'm not sure I fully understand. Do you mean you want to use a GP to predict `p(f_{t}|f_{t-1}, D)`, and then propagate a rollout? This is possible using sampling:
```python
preds_1 = model(test_x)
test_x_samples = preds_1.rsample(torch.Size(n_samples))
model.get_fantasy_model(test_x, test_x_samples)  # Update the model with "fantasy" observations.

# Model is now actually a set of n_samples GPs, where the ith GP has the test data added as training data with the ith sample used as their labels.

preds_2 = model(test_x2)  # Get n_samples predictions for each point in test_x2. E.g., if test_x2 is k x d, then preds_2's mean is n_samples x k.
test_x2_samples = preds_2.rsample(torch.Size(n_samples))  # This will be n_samples x n_samples x k.

# test_x2_samples[i, j, p] is the value of the pth point according to the ith sample, 
# where the model was conditioned on using the jth sample of test_x's function values above
# as the labels.
```

Is this roughly what you are looking for?

**Comment by andrewczgithub** [comment]
hi @jacobrgardner !
Thank you!
I honestly was just wondering, if you had a 1d series vs (n,d) series how would you use the (n,d) series to predict the 1 d series. If these are time series values, Im not really sure how to structure the experiment to predict the 1 d series multiple steps into the future?

Many thanks for your advice.
Best,
Andrew

**Comment by shubhankaray** [comment]
If I may chime in, I think you need a model to tie up the different time series. This can be done in several ways depending on the application, for example, one could use seemingly-unrelated regressions (SUR) for contemporaneous relationships. For GP-based models specifically, I think you can look at the literature on spatio-temporal kriging (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2741335/pdf/nihms127612.pdf). 

Given such a multivariate model, it should be possible to write a (posterior) conditional distribution of 1 time series given the rest...

**Comment by adam-rysanek** [comment]
Hi @andrewczgithub, @jacobrgardner.

I'm trying to work out this solution, as this is also the type of model I'm facing.
I'm trying to understand how to establish predicted values for `f(t-1)` in the independent variable set. 

for example, for my training data, I have pre-established:
``` python
trainX = [x1,x2,x3,t,f(t_1)] # Obviously not the right syntax, but this is just for illustration purposes
trainY = f(t)
```
My test data has a similar structure:
testX = [x1,x2,x3,t,f(t_1)] # Obviously not the right syntax, but this is just for illustration purposes
testY = f(t)

Let's state that a single sample prediction of my model generates `f*(t)=f(x1,x2,x3,t,f(t_1))`

Is the behaviour of `get_fantasy_model` such that, after updating the model with the predicted `f*(t) and its independent variables`), then my next prediction will be as `f*(t+1) = f(x1,x2,x3,t,f*(t))`?

Sorry if my question is not the clear, but it seems to me that how i've structured the model, by including 'f*(t)' in the independent variable set of the model, means that `get_fantasy_model` doesn't update this value for future predictions, unless I'm wrong?

**Comment by gpleiss** [comment]
Hi @adam-rysanek - if I understand your setup correctly, I don't think you want to be using `get_fantasy_model` with the particular setup you have. Because you are supplying `f(t-1)` as an independent variable into the function, you do not need to condition your posterior GP function on that value.

An alternative setup would be not including `f(t_1)` as an independent variable into your function (i.e. your input is `[x1,x2,x3,t]`). In this case, you would want to use `get_fantasy_model`. Hope this helps!

**Comment by gpleiss** [comment]
closing for now - reopen if there's still questions :)

**Comment by q8888620002** [comment]
Building on the previous discussion, is it possible to deal with a streaming time series prediction that given 
```` 
## when t = 1,2,3...n 

L_t =  [[x1_1, x2_1, x3_1 ],
           [x1_2, x2_2, x3_2 ],
             ................
           [x1_n, x2_n, x3_n ]] 

trainX = L_t
trainY =[ x1_n+1, x2_n+1, x3_n+1]
````
So basically is to get G(x1_n+1, x2_n+1, x3_n+1 |   L_t ) which would require training with dynamic dimensional input. I wonder if there is a way to frame this training with Gpytorch.

**Comment by famura** [comment]
Even though this issue is closed, I want to add my 50 cents to it. Partially, in order to see if there are any objections, i.e., if I should adapt my thoughts on this.

In the situation at hand, we assume to have a GP with trained hyperparameters which is supposed to do multiple consecutive/cascaded predictions. Assuming 1-dim data (no loss of generality here), we have for example x[t=0], x[t=1], ..., x[t=T] and want to predict y[t=T], y[t=T+1], y[t=T+H-1], i.e., a horizon on length H. Since we have 1-dim data and assume an autoregressive model, we have the case that our predicted y-values become the input x-values for the next prediction.

To me, it does not make sense to (iteratively) construct a fantasy model to incorporate the predicted values. My argument in non-technical terms is that the GP's "knowledge" is in its (training) data X and Y. Adding predictions to these data sets as if they were observations, i.e., conditioning on our own predictions, artificially reduces the uncertainty at these points.

I think that [this tutorial](https://docs.gpytorch.ai/en/latest/examples/08_Advanced_Usage/SVGP_Model_Updating.html) displays the situation quite well. If you focus on the last plot, in particular the `New Data` points, you will see that the fantasy model is treating the new points as if they were in the training set. However, in our case, I would argue that we do not want this, since we want our consecutive/cascaded predictions to become more uncertain if they deviate from the seen data.

I hope this helps @q8888620002 @andrewczgithub and please let me know if I'm wrong @gpleiss
